protocol/src/lib.rs

```diff
@@ -227,160 +268,184 @@ impl PacketReader {
     pub fn decrypt_contents(
         &mut self,
         ciphertext: &[u8],
         contents: &mut [u8],
         aad: Option<&[u8]>,
-    ) -> Result<(), Error> {
+    ) -> Result<PacketType, Error> {
         let auth = aad.unwrap_or_default();
         // Check minimum size of ciphertext.
-        if ciphertext.len() < TAG_BYTES {
-            return Err(Error::MessageLengthTooSmall);
+        if ciphertext.len() < NUM_TAG_BYTES {
+            return Err(Error::CiphertextTooSmall);
         }
-        let (msg, tag) = ciphertext.split_at(ciphertext.len() - TAG_BYTES);
+        let (msg, tag) = ciphertext.split_at(ciphertext.len() - NUM_TAG_BYTES);
         // Check that the contents buffer is large enough.
         if contents.len() < msg.len() {
             return Err(Error::BufferTooSmall {
                 required_bytes: msg.len(),
             });
         }
         contents[0..msg.len()].copy_from_slice(msg);
-        self.packet_decoding_cipher.decrypt(
+        self.packet_decoding_aead.decrypt(
             auth,
             &mut contents[0..msg.len()],
             tag.try_into().expect("16 byte tag"),
         )?;
 
-        Ok(())
+        Ok(PacketType::from_bytes(contents))
     }
 
-    /// Decrypt the rest of the message from the peer, excluding the 3 length bytes. This method should only be called after
-    /// calling `decrypt_len` on the first three bytes of the buffer.
+    /// Decrypt the packet contents.
     ///
     /// # Arguments
     ///
-    /// - `ciphertext` - The message from the peer.
-    /// - `aad`        - Optional authentication for the peer.
+    /// * `ciphertext` - The packet from the peer excluding the first 3 length bytes. It should contain
+    ///                  the header, contents, and authentication tag.
+    /// * `aad`        - Optional associated authenticated data.
+    ///
+    /// # Returns
+    ///
+    /// A `Result` containing:
+    ///   * `Ok(Some(Vec<u8>))`: The plaintext contents in a byte vector if it is not a decoy packet.
+    ///   * `Err(Error)`: An error that occurred during decryption.
     ///
     /// # Errors
     ///
-    /// Fails if the packet was not decrypted or authenticated properly.
+    /// * `CiphertextTooSmall` - Ciphertext argument does not contain a whole packet.
     #[cfg(feature = "alloc")]
     pub fn decrypt_contents_with_alloc(
         &mut self,
         ciphertext: &[u8],
         aad: Option<&[u8]>,
-    ) -> Result<Vec<u8>, Error> {
-        let mut contents = vec![0u8; ciphertext.len() - TAG_BYTES];
-        self.decrypt_contents(ciphertext, &mut contents, aad)?;
-        Ok(contents)
+    ) -> Result<Option<Vec<u8>>, Error> {
+        let mut contents = vec![0u8; ciphertext.len() - NUM_TAG_BYTES];
+        match self.decrypt_contents(ciphertext, &mut contents, aad)? {
+            PacketType::Decoy => Ok(None),
+            PacketType::Genuine => {
+                // Drop the header byte.
+                contents.remove(0);
```

Comment:
Ah good catch. Removing front elements is such a pain...lot of tradeoffs here and not sure I have thought of something good yet.

Reply:
I think `VecDeque` has an additional byte of overhead but would allow `pop_front` in constant time. Either that or we just give the caller the decoy byte back and let them take a slice `[1..]`

Reply:
`VecDeque` would definitely help the impl, but then returning that is weird for the caller. I am trying to find some uses of it in the ecosystem and am struggling.

Reply:
Few ideas I am kicking around, trying to find the right balance between performance and complexity on the caller. And just to recap, the issue here is the header byte is covered by the auth tag so needs to present in the lower level libs, but is pretty useless for higher level callers.

1. Took a look at the `bytes` crate just for some inspiration since I don't think we would want to add a new dependency. But didn't see anything revolutionary.
2. Use a `VecDeque` instead of a `Vec` for the performance, but requires returning a `VecDeque` (or otherwise cancel out performance gains). I don't think these are used very often, for example can't find a single use in rust-bitcoin. Feels like a strange requirement to place on the caller given the hope is to lower confusion by removing the header byte.
3. Anything involving `Vec` (e.g. `drain`) will at some point require some byte coping, no way around it. Could just leave it for now, but does feel pretty bad.

Does this lead us back to a custom type? Maybe one which owns the underlying vector but supplies a `contents` slice to the caller so they don't have to remember what the header byte is?